Create e2e workflow to upload questions from XL

Zaseem-BIsquared opened the pull request:
= Tested the form manually by uploading XLSX and other file types
- Names of functions and variables are clear
- Comments explain the non-obvious
- Duplication - None

Zaseem-BIsquared committed:
Feature #1 - Upload XL with questions - Create upload form

= Tested the form manually by uploading XLSX and other file types
- Names of functions and variables are clear
- Comments explain the non-obvious
- Duplication - None

Zaseem-BIsquared committed:
Feature #1 - Read the XL and store it in a Database - Create DB

- Tested manually by uploading XLSX and other files and then checked the status message.
- Names of functions and variables are clear
- Comments explain the non-obvious
- Duplication - None

siraj-samsudeen commented on main.py:
```diff
@@ -1,15 +1,39 @@
+        Input(
+            type="file",
+            name="file",
+            accept=".xlsx",  # This restict user to upload only excel file with .xlsx extension
```
this is obvious from the param name. So, comment is not needed.

Zaseem-BIsquared replied:
We removed it.

siraj-samsudeen commented on main.py:
```diff
@@ -1,15 +1,39 @@
+)
+
+
+@rt("/")
```
this kind of formatting differences should NOT exist in commit diff.

Zaseem-BIsquared replied:
We forgot to use a Black formatter before committing, so we changed it to format for every save to avoid this commit diff.

siraj-samsudeen commented on main.py:
```diff
@@ -18,9 +42,36 @@ def get():
     return Title("LMS"), Container(form, Div(id="main"))
 
+
+def standardize_columns(column_name):
```
column_name is a list or string - name of the function is plural and param is singular.

Zaseem-BIsquared replied:
It is a string - we'll change it to singular

siraj-samsudeen commented on main.py:
```diff
@@ -18,9 +42,36 @@ def get():
+    # If there is no answer then the default answer is 'A'
+    df.answers = df.answers.fillna("A") 
+    # Convert null values to empty strings for storing in DB
+    df = df.fillna("")
```
why?

Zaseem-BIsquared replied:
In the first version, when we tried to insert Excel data into the database, we got an error that was hard to understand. We thought it was because of empty values, so we replaced them with empty strings, and the error went away. But now, when we try the same thing, the code works without needing to use fillna(""), Hence we'll remove it.

siraj-samsudeen commented on main.py:
```diff
@@ -18,9 +42,36 @@ def get():
+def convert_binary_to_df(file_content):
+    # Create a BytesIO object to read it as excel
+    excel_data = BytesIO(file_content)
+    df = pd.read_excel(excel_data, dtype=object)
```
why dtype=object?

Zaseem-BIsquared replied:
when we uploaded XL with the column contains numeric values, the pandas treated it as a decimal. so we specified all values as an object when reading, and it displays the same value.

siraj-samsudeen replied:
use realistic test data and do not hard-code types to object now.

siraj-samsudeen commented on main.py:
```diff
@@ -18,9 +42,36 @@ def get():
-    # TODO - File structure is assumed as of now. Hence we are NOT validating it
-    return "File uploaded successfully!"
+async def post(file: UploadFile):
+    if not file.filename.endswith("xlsx"):  
```
why is the comment about file structure deleted?

Zaseem-BIsquared replied:
We thought "file structure" referred to an Excel file, so once we validated that the uploaded file was an Excel file, we removed the comment.

siraj-samsudeen commented on main.py:
```diff
@@ -18,9 +42,36 @@ def get():
-    # TODO - File structure is assumed as of now. Hence we are NOT validating it
-    return "File uploaded successfully!"
+async def post(file: UploadFile):
+    if not file.filename.endswith("xlsx"):  
```
why are you checking for file extension as it is already specified in the accept param in the File Upload?

Zaseem-BIsquared replied:
Even though we specified the accept parameter during testing, we noticed that we could still select the 'All files (.)' option and upload a text file, which caused an error on the backend. To fix this, we added a check for the file type before reading it as an Excel file.

siraj-samsudeen commented on main.py:
```diff
@@ -1,15 +1,39 @@
-app, rt = fast_app(live=True)
 
-@rt('/')
+app, rt, quizs, Quiz = fast_app(
```
use route and quizzes.

Zaseem-BIsquared replied:
we'll update it in the next commit.

siraj-samsudeen commented on main.py:
```diff
@@ -18,9 +42,36 @@ def get():
+    file_content = await file.read()
+    df = convert_binary_to_df(file_content)
+    data = df.to_dict(
+        orient="records"  # This will give each row as a dict value with column name as key
```
you are explaining what the to_dict method will do, not WHY it needs to be called here.

Zaseem-BIsquared replied:
We'll add a comment

siraj-samsudeen reviewed (changes requested):
please install black formatter and set it as the default formatted to avoid formatting-related diffs

Zaseem-BIsquared committed:
Feature #1 - Display the data from db as a table - Add question page

- Tested by comparing the table with uploaded XL.
- Names of functions and variables are clear
- Comments explain the non-obvious
- Duplication - None

Zaseem-BIsquared committed:
Feature #1 - Download the question as XL - Add Export Button

- Tested by manually uploading and exporting to ensure all content is readable.
- Names of functions and variables are clear
- Comments explain the non-obvious
- Duplication - None

Zaseem-BIsquared committed:
Add Playwright E2E test for file upload, display, and export functionality #1
